File: android/src/main/java/AppEventsPlugin.kt
package wang.tato.tauri_plugin_app_events

import android.app.Activity
import app.tauri.Logger
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Channel
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@InvokeArg
class SetEventHandlerArgs {
    lateinit var handler: Channel
}

@TauriPlugin
class AppEventsPlugin(private val activity: Activity) : Plugin(activity) {
    private var resumeChannel: Channel? = null
    private var pauseChannel: Channel? = null

    override fun onResume() {
        Logger.debug("TAURI_PLUGIN_APP_EVENTS", "onResume")
        val event = JSObject()
        trigger("resume", event)
        resumeChannel?.send(event)
        super.onResume()
    }

    @Command
    fun setResumeHandler(invoke: Invoke) {
        Logger.debug("TAURI_PLUGIN_APP_EVENTS","setResumeHandler")
        val args = invoke.parseArgs(SetEventHandlerArgs::class.java)
        resumeChannel = args.handler
        invoke.resolve()
    }

    override fun onPause() {
        Logger.debug("TAURI_PLUGIN_APP_EVENTS", "onPause")
        val event = JSObject()
        trigger("pause", event)
        pauseChannel?.send(event)
        super.onPause()
    }

    @Command
    fun setPauseHandler(invoke: Invoke) {
        Logger.debug("TAURI_PLUGIN_APP_EVENTS","setPauseHandler")
        val args = invoke.parseArgs(SetEventHandlerArgs::class.java)
        pauseChannel = args.handler
        invoke.resolve()
    }

}

File: ios/tauri-plugin-app-events/AppEventsPlugin.swift
//
//  AppEventsPlugin.swift
//  tauri-plugin-app-events
//
//  Created by wtto on 2024/10/28.
//

import OSLog
import SwiftRs
import Tauri
import UIKit
import WebKit

let log = OSLog(subsystem: "com.tauri.dev", category: "plugin.app.events")

class SetEventHandlerArgs: Decodable {
  let handler: Channel
}

class AppEvetnsPlugin: Plugin {
  private var resumeChannel: Channel? = nil
  private var pauseChannel: Channel? = nil

  override init() {
    super.init()
    NotificationCenter.default.addObserver(self, selector: #selector(applicationDidBecomeActive), name: UIApplication.didBecomeActiveNotification, object: nil)
    NotificationCenter.default.addObserver(self, selector: #selector(applicationDidEnterBackground), name: UIApplication.didEnterBackgroundNotification, object: nil)
  }

  @objc func applicationDidBecomeActive(notification: NSNotification) {
    os_log(.debug, log: log, "Application Did Become Active")
    trigger("resume", data: JSObject())
    resumeChannel?.send(JSObject())
  }

  @objc func applicationDidEnterBackground(notification: NSNotification) {
    os_log(.debug, log: log, "Application Did Enter Background")
    trigger("pause", data: JSObject())
    pauseChannel?.send(JSObject())
  }

  @objc public func setResumeHandler(_ invoke: Invoke) throws {
    os_log(.debug, log: log, "setResumeHandler")
    let args = try invoke.parseArgs(SetEventHandlerArgs.self)
    resumeChannel = args.handler
    invoke.resolve()
  }

  @objc public func setPauseHandler(_ invoke: Invoke) throws {
    os_log(.debug, log: log, "setPauseHandler")
    let args = try invoke.parseArgs(SetEventHandlerArgs.self)
    pauseChannel = args.handler
    invoke.resolve()
  }
}

@_cdecl("init_plugin_app_events")
func initPlugin() -> Plugin {
  return AppEvetnsPlugin()
}
